README.md
# DeepShift
## Towards a Multiplication-Less Neural Network

This is a project that aims to replace multiplications in a neural networks with bitwise shift (and sign change).

### Results So Far
#### Converting all Layers
Converting all `Conv2D` layers to `Conv2DShift` and all `Dense` layers to `DenseShift`:

| Model | Dataset | Original Version | DeepShift Version | 
|-------| ------- | -------------------------- | ----------------------------- |
| ResNet20 | CIFAR10 | 92.16% | <del>77.02%<del> |
| ResNet32 | CIFAR10 | 92.46% | <del>85.54%<del> |
| ResNet44 | CIFAR10 | 92.50% | <del>85.34%<del> |
| ResNet50 | CIFAR10 | 91.93% | <del>86.23%<del> |
| ResNet56 | CIFAR10 | 92.71% |  <del>86.30%<del> |
| ResNet110 | CIFAR10 | 92.65% | 67.86% |

#### Converting Some Layers
Converting only the last `N_shift` convolution layers (as well as the last fully connected layer) to shift layers.
The table below shows the validation accuracy results for ResNet50 on CIFAR10:

| # Conv2D Layers | # Conv2DShift Layers | Accuracy |
| --------------- | -------------------- | -------- |
| 51 | 0 | 91.93% |
| 39 | 12 | 92.090% <del>92.14%<del> |
| 29 | 22 | 87.02% <del>91.65%<del> |
| 19 | 32 | 82.81%<del>88.83%<del> |
| 9 | 42 | <del>88.12%<del> |
| 0 | 51 | <del>86.23%<del> |

### Codewalk Through
* `keras`: directory containing implementation, tests, and saved models using Keras
    * `shift_layer.py`: definition of `DenseShift` class that inherits from Keras' `Layer` class and modifies it to implement fully connected (a.k.a Dense) layer as multiplications with integer powers of 2 (mathematically equialvent to bitwaise shift) and  integer powers of -1 (mathematically equivalent to sign change or sign keep). 
    * `convolutional_shift.py`: definition of `Conv2DShift` class that inherits from Keras` `Conv2D` layer to implement convolution as multiplications with integer powers of 2 (mathematically equialvent to bitwaise shift) and  integer powers of -1 (mathematically equivalent to sign change or sign keep).
    * `mnist_deepshift.py`: example script to classify MNIST dataset using a small fully connected model.
    * `cifar10_resshift.py`: modify different depth versions of ResNet50 to use shift convolutions instead of multiplication convolutions.
    * `selected_models`: directory containing saved Keras model files and training results

### TODOs
Currently, both trainining and inference are actually done by multiplying by a power of 2 and by a power of -1.
So we need to:
- Implement the inference on fixed float data using bitwise shift (instead of multiplying by a power of 2) and bitwise XOR (instead of negating), and verify that its speed on CPU will be faster than multiplication.
- Implement training on fixed float data using bitwaise shift and bitwise XOR, and verify that its speed on CPU will be faster than multiplication.
- Implement an FPGA prototype to execute bitwise shift and XOR  in parallel.
- Investigate different methods or hyperparameter tuning to enhance the accuracy results.

### Running the Code
1. Clone the repo:
```
git clone http://rnd-gitlab-ca-y.huawei.com/mindspore/deepshift/ DeepShift
```
2. Change directory
```
cd DeepShift
```
3. Create virtual environment: 
```
virtualenv venv
```
4. (Needs to be done every time you run code) Source the environment:
```
source venv/bin/activate
```
5. Install required packages and build the spfpm package for fixed point
```
pip install -r requirements.txt
cd spfpm
make
```
6. cd into `keras` directroy:
```
cd keras
```
7. Run the MNIST test:
```
python mnist_deepshift.py
```
8. Run the ResNet test. You can check for various options to pass:
```
python cifar10_resshift.py --help
```
and then run with the script with default options or pass the options you prefer:
```
python cifar10_resshift.py
```


spfpm/ChangeLog
ChangeLog for Simple Python Fixed-Point Module

12May18	- Added signed/twos-complement option to FXnum.toBinaryString() method

03May18	- Added 'precision' option to FXnum.toDecimalString() method
	Improved estimation of number of digits in decimal representation

28Apr18	- Replaced asin() formula for pi with Bailey-Borwein-Plouffe series

16Apr18	- *** spfpm-1.4 released

13Apr18	- Added method for printing FXnum as binary/octal/hexadecimal string

11Apr18	- Added shortcut method to FXfamily for creating FXnum instances

24Feb18	- *** spfpm-1.3 released

24Jun17	- Improved convergence of asin() series.
	Replaced atan() computation of pi with asin().

29May17	- Neatened asin() and acos() methods, avoiding use of atan().

28May17	- Improved efficiency of computation of log(2)
	Converted README.txt into markdown

13May17	- Added optimized FXnum constructor to improve speed of arithmetic
	Added command-line switches for demonstration script
	Added plot of Pi accuracy to demonstration script

06May17	- *** spfpm-1.2 released

05May17	- removed need to have same FXfamily object when combining FXnums
	- added '__slots__' specifier to FXnum fields

15Jul14	- *** spfpm-1.1 released

15Jul14	- added left/right shift operators
	improved initialization of FXnum objects from output of repr()

14Jul14	- replaced various getter methods in FXfamily with properties
	added improved expressions for computing exp(1) and log(2)
	updated graph-based demos to use Matplotlib in place of Gnuplot

19May09	- *** spfpm-1.0.1 released

20Feb09	- patched FXnum constructor to check for overflows

31Jan09	- *** spfpm-1.0 released

31Jan09	- improved accuracy of pi,log(2) etc used in trig,exp functions

06Dec08	- ported to Python-3.0

11Apr08	- *** spfpm-0.6 released

09Apr08	- added validator in FXfamily to allow limited range of integer-part

07Oct07	- added basic FLOPS calculator to demo.py

27May07	- *** spfpm-0.5 released

26May07	- minor adjustments to ensure test-cases run under Python-2.2
	added unit-test for repr()

21May07	- neatened module importing in unit-test routines

23Apr07	- removed use of 'isinstance' from FXfamily & (almost all of) FXnum

11Feb07	- added improved asin()/acos() methods based on 1-cos2t transformations

04Feb07	- removed (deprecated) FPnum wrappers

03Feb07	- *** spfpm-0.4 released

03Feb07	- patched rounding on cast to int/long to match behaviour of float-casts
	added unit-test for sqrt() method
	expanded set of exception classes, including domain-error etc
	added unit-test for asin() & acos() methods

02Feb07	- added demo/test/clean Makefile targets

01Feb07	- added setup.py distutils-script

31Jan07	- added basic asin() & acos() methods

30Jan07	- replaced _rawsincos() method with dual-mode _rawQsine() method
	added sin() & cos() methods

29Jan07	- patched spurious family-conversion in __neg__() operator

27Jan07	- extended discusions + examples in documentation
	added doctest.testmod() test for documentation
	added __abs__, __pos__, __hash__() methods
	changed default distribution-format to .zip in place of .tar.gz

22Jan07	- neatened internal usage of truncating division
	wired-together __div__ & __truediv__ operators

21Jan07	- *** spfpm-0.3 released

21Jan06	- added backwards-compatibility SetFraction method to FPnum wrapper
	renamed arctan() method to atan() to agree with math-module convention

20Jan06	- added unit-test for resolution-conversions
	added arctan & _rawarctan() methods
	adjusted sincos() method to reduce angles moduli Pi rather than halving
	added GetPi() method to FXfamily
	adjusted FXfamily to use greater bit-count when finding cached constants

19Jan06	- changed 'FPnum' to 'FXnum' to avoid abbreviation-clash with 'Floating Point'
	added FPnum backward-compatibility wrapper

18Jan06	- added unit-test for __str__ method

17Jan06	- converted resolution information to shareable FPfamily object

16Jan06	- added test for __nonzero__ method & patched reference to 'other'
	removed __iadd__, __isub__ methods to make immutable object
	added unit-test for immutability

15Jan07	- *** spfpm-0.2 released

13Jan07	- added basic Makefile for building release packages
	added unit-tests for pow() and sincos() methods

08Jan07	- added logarithm methods

03Jan07	- added basic unit-testing class for arithmetic operations

02Jan07	- converted RCS repository into SVN

26Dec06	- *** prototype of spfpm-0.1 released


spfpm/demo.py
#!/usr/bin/python3
# Demonstration of Simple Python Fixed-Point Module
# (C)Copyright 2006-2018, RW Penney

from __future__ import print_function
import argparse, time
from collections import OrderedDict
try:
    import matplotlib, numpy
    import matplotlib.pyplot as plt
    HAVE_MATPLOTLIB = True
except ImportError:
    HAVE_MATPLOTLIB = False

import FixedPoint


def basicDemo():
    """Basic demonstration of roots & exponents at various accuracies"""

    for resolution in [8, 32, 80, 274]:
        family = FixedPoint.FXfamily(resolution)
        val = 2

        print('=== {0} bits ==='.format(resolution))
        rt = family(val).sqrt()
        print('sqrt(' + str(val) + ')~ ' + str(rt))
        print('sqrt(' + str(val) + ')^2 ~ ' + str(rt * rt))
        print('exp(1) ~ ' + str(family.exp1))
        print()


def overflowDemo():
    """Illustrate how finite range limits calculation of exponents"""

    res = 20
    print('=== {0}-bit fractional part ==='.format(res))
    for intsize in [4, 8, 16, 32]:
        family = FixedPoint.FXfamily(res, intsize)
        x = family(0.0)
        step = 0.1
        while True:
            try:
                ex = x.exp()
            except FixedPoint.FXoverflowError:
                print('{0:2d}-bit integer part: exp(x) overflows near x={1:.3g}'.format(intsize, float(x)))
                break
            x += step
    print()


def speedDemo():
    """calculate indicative speed of floating-point operations"""

    print('=== speed test ===')
    for res, count in [ (16, 10000), (32, 10000),
                        (64, 10000), (128, 10000),
                        (256, 10000), (512, 10000) ]:
        fam = FixedPoint.FXfamily(res)
        x = fam(0.5)
        lmb = fam(3.6)
        one = fam(1.0)
        t0 = time.clock()
        for i in range(count):
            # use logistic-map in chaotic region:
            x = lmb * x * (one - x)
        t1 = time.clock()
        ops = count * 3
        Dt = t1 - t0
        print('{0} {1}-bit arithmetic operations in {2:.2f}s ~ {3:.2g} FLOPS' \
                .format(ops, res, Dt, (ops / Dt)))

    for res, count in [ (4, 10000), (8, 10000), (12, 10000),
                        (24, 10000), (48, 10000), (128, 10000),
                        (512, 10000) ]:
        fam = FixedPoint.FXfamily(res, 4)
        x = fam(2)
        t0 = time.clock()
        for i in range(count):
            y = x.sqrt()
        t1 = time.clock()
        Dt = (t1 - t0)
        print('{} {}-bit square-roots in {:.3g}s ~ {:.3g}/ms' \
                .format(count, res, Dt, count*1e-3/Dt))


def printBaseDemo():
    res = 60
    pi = FixedPoint.FXfamily(res).pi

    print('==== Pi at {}-bit resolution ===='.format(res))
    print('decimal: {}'.format(pi.toDecimalString()))
    print('binary: {}'.format(pi.toBinaryString()))
    print('octal: {}'.format(pi.toBinaryString(3)))
    print('hex: {}'.format(pi.toBinaryString(4)))


def piPlot():
    """Plot graph of approximations to Pi"""

    b_min, b_max = 8, 25
    pi_true = FixedPoint.FXfamily(b_max + 40).pi
    pipoints = []
    for res in range(b_min, b_max+1):
        val = 4 * FixedPoint.FXnum(1, FixedPoint.FXfamily(res)).atan()
        pipoints.append([res, val])
    pipoints = numpy.array(pipoints)
    truepoints = numpy.array([[b_min, pi_true], [b_max, pi_true]])

    plt.xlabel('bits')
    plt.ylabel(r'$4 \tan^{-1} 1$')
    plt.xlim([b_min, b_max])
    plt.ylim([3.13, 3.16])
    plt.grid(True)
    for arr, style in ((truepoints, '--'), (pipoints, '-')):
        plt.plot(arr[:,0], arr[:,1], ls=style)
    plt.show()


class ConstAccuracyPlot(object):
    """Plot graph of fractional bits wasted due to accumulated roundoff."""

    @classmethod
    def calcConst(cls, fam):
        return FXnum(1, fam)

    @classmethod
    def getConst(cls, fam):
        return fam.unity

    @classmethod
    def getLabels(cls):
        return ('1', '1')

    @staticmethod
    def lostbits(x, x_acc):
        fam_acc = x_acc.family
        eps = (FixedPoint.FXnum(x, fam_acc) - x_acc)
        return (abs(eps).log() / fam_acc.log2
                    + x.family.resolution)

    @classmethod
    def draw(cls):
        losses = []
        for bits in range(4, 500, 4):
            fam_acc = FixedPoint.FXfamily(bits + 40)
            fam = FixedPoint.FXfamily(bits)
            const_true = cls.getConst(fam_acc)
            const_family = cls.getConst(fam)
            approx = cls.calcConst(fam)
            losses.append((bits, cls.lostbits(approx, const_true),
                                 cls.lostbits(const_family, const_true)))
        losses = numpy.array(losses)

        plt.xlabel('resolution bits')
        plt.ylabel('error bits')
        plt.grid(True)
        lab_approx, lab_family = cls.getLabels()
        plt.plot(losses[:,0], losses[:,1], label=lab_approx)
        plt.plot(losses[:,0], losses[:,2], label=lab_family)
        plt.legend(loc='best', fontsize='small')
        plt.show()


class PiAccuracyPlot(ConstAccuracyPlot):
    @classmethod
    def calcConst(cls, fam):
        return 6 * FixedPoint.FXnum(0.5, fam).asin()

    @classmethod
    def getConst(cls, fam):
        return fam.pi

    @classmethod
    def getLabels(cls):
        return (r'$6 \sin^{-1} \frac{1}{2}$', r'$\pi_{family}$')


class Log2AccuracyPlot(ConstAccuracyPlot):
    @classmethod
    def calcConst(cls, fam):
        return (2 * FixedPoint.FXnum(3, fam).log()
                - (FixedPoint.FXnum(9, fam) / 8).log()) / 3

    @classmethod
    def getConst(cls, fam):
        return fam.log2

    @classmethod
    def getLabels(cls):
        return (r'$(\log9 - \log(9/8))/3$', r'$\log2_{family}$')


def main():
    demos = OrderedDict([
        ('basic',       basicDemo),
        ('overflow',    overflowDemo),
        ('speed',       speedDemo),
        ('printing',    printBaseDemo) ])
    if HAVE_MATPLOTLIB:
        demos['piplot'] = piPlot
        demos['piaccplot'] = PiAccuracyPlot.draw
        demos['log2accplot'] = Log2AccuracyPlot.draw

    parser = argparse.ArgumentParser(
                description='Rudimentary demonstrations of'
                            ' the Simple Python Fixed-point Module (SPFPM)')
    parser.add_argument('-a', '--all', action='store_true')
    parser.add_argument('demos', nargs='*',
            help='Demo applications ({})'.format('/'.join(demos.keys())))
    args = parser.parse_args()

    if args.all or not args.demos:
        args.demos = list(demos.keys())

    for demoname in args.demos:
        demos[demoname]()


if __name__ == "__main__":
    main()

# vim: set ts=4 sw=4 et:
